[JENKINS-68933] Update to Jetty 10, with WebSocket

jglick opened the pull request:
Subsumes #6780; reverts #6781, reapplying #6694.

See [JENKINS-68933](https://issues.jenkins.io/browse/JENKINS-68933).

Development loop:

```bash
mvnd -Pquick-build install -pl websocket/spi,websocket/jetty9,websocket/jetty10,core,war
mvnd -pl test test -Dtest=JNLPLauncherRealTest\#webSocket
```

### Proposed changelog entries

* Upgrading the bundled Jetty to version 10, via Winstone 6.0.

### Proposed upgrade guidelines

N/A

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6785"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

jglick committed:
[JENKINS-68933] Verify basic WebSocket connectivity in `RealJenkinsRule`

jglick committed:
Removing use of reflection to access Jetty WebSocket types

jglick committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins into ws-JENKINS-68933

jglick committed:
Removing `javax.servlet-api` from BOM

jglick committed:
Tweaking packaging

jglick committed:
Revert "Revert "[JENKINS-68694] Winstone 6.0: Upgrade Jetty from 9.4.46.v20220331 to 10.0.11" (#6781)"

This reverts commit 4ba9c08901cc7ca3ae61d79f08f83411a481b78a.

jglick committed:
Currently broken draft of Jetty 10 provider

jglick commented on websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:
```diff
@@ -0,0 +1,163 @@
+
+    private synchronized void init(HttpServletRequest req) throws Exception {
+        if (container == null) {
+            // TODO JettyWebSocketServletContainerInitializer.configure(context, null) but where do we get the ServletContextHandler?
```
https://github.com/eclipse/jetty.project/pull/5874 presumes a different scenario.

jglick replied:
Currently I am stumped here. @olamy any idea at all?

basil replied:
> Currently I am stumped here.

@jglick I am disappointed you did not ask me. The instance of `ServletContextHandler` in Winstone is the `org.eclipse.jetty.webapp.WebAppContext` created in `winstone.HostConfiguration#create`. The call to `org.eclipse.jetty.websocket.server.config.JettyWebSocketServletContainerInitializer#configure` has to be done before the call to `org.eclipse.jetty.server.Server#start` so we might as well do it immediately after creating the `org.eclipse.jetty.webapp.WebAppContext` in `winstone.HostConfiguration#create`. With that in place, the call to `org.eclipse.jetty.server.Server#start` will do two important things: invoke `org.eclipse.jetty.websocket.server.config.JettyWebSocketConfiguration`'s constructor to expose the relevant classes, and invoke `org.eclipse.jetty.websocket.server.JettyWebSocketServerContainer#ensureContainer`, which creates the container and stashes it in the servlet context with attribute name `WebSocketContainer.class.getName()`, from which it can be fished out in the application.

So to summarize, patch Winstone with

```java
diff --git a/src/main/java/winstone/HostConfiguration.java b/src/main/java/winstone/HostConfiguration.java
index be7a223..21e9dab 100644
--- a/src/main/java/winstone/HostConfiguration.java
+++ b/src/main/java/winstone/HostConfiguration.java
@@ -14,6 +14,7 @@ import org.eclipse.jetty.server.Server;
 import org.eclipse.jetty.server.handler.ContextHandlerCollection;
 import org.eclipse.jetty.server.handler.RequestLogHandler;
 import org.eclipse.jetty.webapp.WebAppContext;
+import org.eclipse.jetty.websocket.server.config.JettyWebSocketServletContainerInitializer;
 import winstone.cmdline.Option;
 
 import javax.servlet.SessionTrackingMode;
@@ -181,6 +182,7 @@ public class HostConfiguration {
                 getSessionHandler().getSessionCache().setEvictionPolicy( sessionEviction );
             }
         };
+        JettyWebSocketServletContainerInitializer.configure(wac, null);
         wac.getSecurityHandler().setLoginService(loginService);
         wac.setThrowUnavailableOnStartupException(true);    // if boot fails, abort the process instead of letting empty Jetty run
         wac.setMimeTypes(mimeTypes);
```

and then amend this PR with

```java
diff --git a/websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java b/websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
index af8b6125a1..694b9f6297 100644
--- a/websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
+++ b/websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
@@ -33,6 +33,7 @@ import java.util.concurrent.Future;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.WebSocketContainer;
 import org.eclipse.jetty.websocket.api.WebSocketListener;
 import org.eclipse.jetty.websocket.api.WriteCallback;
 import org.eclipse.jetty.websocket.server.JettyServerUpgradeRequest;
@@ -59,11 +60,7 @@ public class Jetty10Provider implements Provider {
 
     private synchronized void init(HttpServletRequest req) throws Exception {
         if (container == null) {
-            // TODO JettyWebSocketServletContainerInitializer.configure(context, null) but where do we get the ServletContextHandler?
-            container = JettyWebSocketServerContainer.getContainer(req.getServletContext());
-            if (container == null) {
-                throw new IllegalStateException("not initialized");
-            }
+            container = (JettyWebSocketServerContainer) req.getServletContext().getAttribute(WebSocketContainer.class.getName());
         }
     }
```

I just tested this and it seems to work fine.

basil replied:
BTW not sure if anybody has any strong feelings about whether that call to `JettyWebSocketServletContainerInitializer.configure` in Winstone should be done unconditionally or should be opt-in/opt-out via CLI and/or system property. If opt-in is desired, Jenkins could opt-in via `executable.Main` in the `war/`module in core (previously the `extras-executable-war` repository). I doubt anyone would ever have a need to turn this off, so my initial take is that it should just be enabled unconditionally for simplicity, but I do not feel strongly one way or another.

basil replied:
(See also https://github.com/jenkinsci/jenkins-test-harness/pull/453/commits/7a2234286b681c5c4831295ef992a332e43e5f97)

jglick replied:
Many thanks! I must have been tired and did not think to look for _subtypes_ of `ServletContextHandler`.

basil committed:
Checkstyle

basil committed:
SpotBugs

basil committed:
Enforcer

basil reviewed (commented):
I was exploring this logic earlier today and came to a similar conclusion about `isUpgradeRequest`. I had been thinking about inlining it, but I guess retaining only the first check as you did here is probably good enough.

This looks good pending a few remaining tasks:

* Merge in the static analysis fixes I just made to #6780 because otherwise this won't build in CI
* Probably need to add Enforcer excludes for the new modules as well to get Enforcer to pass on the `war/` module
* Add the call to `JettyWebSocketServletContainerInitializer.configure` in Winstone per https://github.com/jenkinsci/jenkins/pull/6785#discussion_r914305175
* Once the CI build is passing, I would personally like to run another build with the Jetty 10 based `JenkinsRule` from https://github.com/jenkinsci/jenkins-test-harness/pull/453 and PCT/BOM testing with the same. It should not turn up anything new, but I would rather be safe than sorry. With passing core and BOM/PCT suites with a Jetty 10 based `JenkinsRule` I would feel very confident about this change.

Happy to help with these tasks, so let me know how we can best drive this forward together.

jglick committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins into ws-JENKINS-68933

jglick committed:
Skip Javadoc on new modules

jglick committed:
Merge branch 'ws-JENKINS-68933' into ws-JENKINS-68933-jetty-10

jglick committed:
Skip Javadoc on new module

jglick committed:
`enforceBytecodeVersion` exclude

jglick committed:
Also exclude `Jetty10Provider` from SpotBugs

jglick committed:
Checkstyle

jglick committed:
Pick up https://github.com/jenkinsci/winstone/pull/240

jglick commented:
BTW regarding `isUpgradeRequest`: https://github.com/eclipse/jetty.project/blob/475d0512a57da08d0e24ecf6a2d01cfc72e11548/jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/ProgrammaticWebSocketUpgradeTest.java#L88-L92 does not bother, and https://github.com/eclipse/jetty.project/blob/475d0512a57da08d0e24ecf6a2d01cfc72e11548/documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/websocket/WebSocketServerDocs.java#L289-L292 is less than helpful.

jglick commented:
> help with these tasks

I think the first three are now covered, so in principle this PR should now be ready as far as WebSocket support is concerned as soon as the Winstone change is released. If you want to set up a separate draft PR for testing against https://github.com/jenkinsci/jenkins-test-harness/pull/453 that would certainly increase our confidence.

(There are apparently other regressions like https://github.com/jenkinsci/jenkins/pull/6694#issuecomment-1174668506 and https://github.com/jenkinsci/jenkins/pull/6694#issuecomment-1173721597 which could be addressed as well. I was not planning to look into those, so if you want to look into them, it may make more sense to file a fresh PR subsuming this one which is truly production-ready. My interest was in restoring WS functionality, especially since I was responsible for the reflection tech debt that made it tricky to understand where Jetty 10 was not working; I guess it did not even occur to me at the time that Jetty would change their APIs incompatibly.)

jglick commented:
Interactively verified `wsecho/` (in both text and binary modes, incl. ping visible after 30s with `websocat -vv`); Pipeline builds on a WS agent; CLI in `-webSocket` mode.

jglick committed:
Treat `ClosedChannelException` specially.

Otherwise closing connections sometimes produces stack traces

```
WARNING	j.websocket.WebSocketSession#error: unhandled WebSocket service error
java.nio.channels.ClosedChannelException
	at org.eclipse.jetty.websocket.core.internal.WebSocketSessionState.onEof(WebSocketSessionState.java:169)
	at org.eclipse.jetty.websocket.core.internal.WebSocketCoreSession.onEof(WebSocketCoreSession.java:253)
	at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.fillAndParse(WebSocketConnection.java:482)
	at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.onFillable(WebSocketConnection.java:340)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:319)
	at …
```

or

```
WARNING	j.agents.WebSocketAgents$Session#error
java.nio.channels.ClosedChannelException
	at org.eclipse.jetty.websocket.core.internal.WebSocketSessionState.onEof(WebSocketSessionState.java:169)
	at org.eclipse.jetty.websocket.core.internal.WebSocketCoreSession.onEof(WebSocketCoreSession.java:253)
	at …
```

though I am not sure if that is actually new behavior.

jglick committed:
Treat `ClosedChannelException` specially.

Otherwise closing connections sometimes produces stack traces

```
WARNING	j.websocket.WebSocketSession#error: unhandled WebSocket service error
java.nio.channels.ClosedChannelException
	at org.eclipse.jetty.websocket.core.internal.WebSocketSessionState.onEof(WebSocketSessionState.java:169)
	at org.eclipse.jetty.websocket.core.internal.WebSocketCoreSession.onEof(WebSocketCoreSession.java:253)
	at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.fillAndParse(WebSocketConnection.java:482)
	at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.onFillable(WebSocketConnection.java:340)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:319)
	at …
```

or

```
WARNING	j.agents.WebSocketAgents$Session#error
java.nio.channels.ClosedChannelException
	at org.eclipse.jetty.websocket.core.internal.WebSocketSessionState.onEof(WebSocketSessionState.java:169)
	at org.eclipse.jetty.websocket.core.internal.WebSocketCoreSession.onEof(WebSocketCoreSession.java:253)
	at …
```

though I am not sure if that is actually new behavior.

jglick committed:
https://github.com/jenkinsci/winstone/pull/240 released

jglick committed:
Do not bundle `javax.servlet-api`: https://github.com/jenkinsci/jenkins/pull/6780#discussion_r915086353

jglick committed:
Do not bundle `javax.servlet-api`: https://github.com/jenkinsci/jenkins/pull/6780#discussion_r915086353

jglick commented:
FYI I was topologically merging in the upstream branch, which Git tracks better than cherry-picks. (Until you squash-merge the upstream PR. One reason I prefer to use true merges.)

basil commented on war/pom.xml:
```diff
@@ -96,6 +96,16 @@ THE SOFTWARE.
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci.main</groupId>
+      <artifactId>websocket-jetty10</artifactId>
+      <version>${project.version}</version>
+    </dependency>
```
This seems to make all of the `JenkinsRule`-based tests start failing with:

```
[ERROR] jenkins.slaves.UnsupportedRemotingAgentEscapeHatchTest.shouldBeAbleToConnectAgentWithUnsupportedVersionWithEscapeHatch  Time elapsed: 0.06 s  <<< ERROR!
java.lang.NoSuchFieldError: CACHE
        at org.eclipse.jetty.http2.hpack.HpackContext.<clinit>(HpackContext.java:126)
        at org.eclipse.jetty.http2.hpack.HpackFieldPreEncoder.getEncodedField(HpackFieldPreEncoder.java:68)
        at org.eclipse.jetty.http.PreEncodedHttpField.<init>(PreEncodedHttpField.java:102)
        at org.eclipse.jetty.http.PreEncodedHttpField.<init>(PreEncodedHttpField.java:108)
        at org.eclipse.jetty.http2.hpack.HpackEncoder.<clinit>(HpackEncoder.java:76)
        at org.eclipse.jetty.http2.hpack.HpackFieldPreEncoder.getEncodedField(HpackFieldPreEncoder.java:38)
        at org.eclipse.jetty.http.PreEncodedHttpField.<init>(PreEncodedHttpField.java:102)
        at org.eclipse.jetty.http.PreEncodedHttpField.<init>(PreEncodedHttpField.java:108)
        at org.eclipse.jetty.http.MimeTypes$Type.<init>(MimeTypes.java:98)
        at org.eclipse.jetty.http.MimeTypes$Type.<clinit>(MimeTypes.java:56)
        at org.eclipse.jetty.http.MimeTypes.<clinit>(MimeTypes.java:175)
        at org.jvnet.hudson.test.JenkinsRule.<clinit>(JenkinsRule.java:2967)
        at jenkins.slaves.UnsupportedRemotingAgentEscapeHatchTest.<init>(UnsupportedRemotingAgentEscapeHatchTest.java:32)
        […]
```

I got the problem to go away with:

```xml
diff --git a/test/pom.xml b/test/pom.xml
index e10e9ea835..1e69250152 100644
--- a/test/pom.xml
+++ b/test/pom.xml
@@ -117,6 +117,12 @@ THE SOFTWARE.
       <version>${project.version}</version>
       <type>executable-war</type>
       <scope>test</scope>
+      <exclusions>
+        <exclusion>
+          <groupId>org.jenkins-ci.main</groupId>
+          <artifactId>websocket-jetty10</artifactId>
+        </exclusion>
+      </exclusions>
     </dependency>
     <dependency>
       <groupId>io.jenkins.plugins</groupId>
```

With that fix the `JenkinsRule` and `RealJenkinsRule` tests both seem to work. I think it makes sense to put that in until the test harness moves to Jetty 10.

jglick replied:
Not sure that is quite right. IIUC now the `test` module is loading Jetty from `websocket-jetty9` rather than `jenkins-test-harness`? We want `websocket-jetty9` to be in the test classpath (actually `websocket-jetty10` itself should be fine as well); we just do not want its transitive dependencies to also be in the test classpath. Maybe they should be using `optional` or `provided` on their deps.

basil replied:
Are you going to try that out or are you asking me to?

basil replied:
Ah I see you just pushed some commits. Yeah that does seem to fix it for me. Here's hoping for a green PR build. :crossed_fingers:

basil reviewed (commented):
> I was not planning to look into those, so if you want to look into them, it may make more sense to file a fresh PR subsuming this one which is truly production-ready.

Sure, let's get a green CI build for #6785 and let's get #6780 merged and then I can file a new PR for the final reintegration. Many thanks for the collaboration thus far. I plan to do two sets of core PRs (one with regular test harness and one with Jetty 10 test harness) and two corresponding BOM/PCT runs.

basil committed:
Exclude Jetty 10 from test project

jglick committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins into ws-JENKINS-68933

jglick committed:
Do not propagate `winstone` dep to `test`: https://github.com/jenkinsci/jenkins/pull/6785#discussion_r915259132

jglick committed:
Merge branch 'ws-JENKINS-68933' into ws-JENKINS-68933-jetty-10

jglick committed:
Do not propagate `winstone` dep to `test`: https://github.com/jenkinsci/jenkins/pull/6785#discussion_r915259132

jglick committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins into ws-JENKINS-68933-jetty-10
